File: src/test/java/ca/joelathiessen/kaly2/tests/pc/unit/featuredetector/SplitAndMergeTest.java
package ca.joelathiessen.kaly2.tests.pc.unit.featuredetector;

import ca.joelathiessen.kaly2.featuredetector.Kaly2Feature;
import ca.joelathiessen.kaly2.featuredetector.SplitAndMerge;
import ca.joelathiessen.kaly2.subconscious.Measurement;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.runners.MockitoJUnitRunner;

import java.util.ArrayList;
import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

@RunWith(MockitoJUnitRunner.class)
public class SplitAndMergeTest {
    private static double EPSILON = 0.001;

    private Measurement makeMeasFromXY(float x, float y) {
        return new Measurement((float) Math.sqrt(x * x + y * y), (float) Math.atan2(y, x), null, 0);
    }

    @Test
    public void testFourPointSquare() {
        SplitAndMerge merge = new SplitAndMerge();

        ArrayList<Measurement> measurements = new ArrayList<>();
        measurements.add(makeMeasFromXY(0, 0));
        measurements.add(makeMeasFromXY(0, 1));
        measurements.add(makeMeasFromXY(1, 1));
        measurements.add(makeMeasFromXY(1, 0));

        List<? extends Kaly2Feature> features = merge.getFeatures(measurements);

        assertTrue(features.size() == measurements.size());

        assertEquals(features.get(0).getX(), 0, EPSILON);
        assertEquals(features.get(0).getY(), 0, EPSILON);

        assertEquals(features.get(1).getX(), 0, EPSILON);
        assertEquals(features.get(1).getY(), 1, EPSILON);

        assertEquals(features.get(2).getX(), 1, EPSILON);
        assertEquals(features.get(2).getY(), 1, EPSILON);

        assertEquals(features.get(3).getX(), 1, EPSILON);
        assertEquals(features.get(3).getY(), 0, EPSILON);
    }

    @Test
    public void testFourPointLine() {
        SplitAndMerge merge = new SplitAndMerge();

        ArrayList<Measurement> measurements = new ArrayList<>();
        measurements.add(makeMeasFromXY(0, 0));
        measurements.add(makeMeasFromXY(0, 1));
        measurements.add(makeMeasFromXY(0, 2));
        measurements.add(makeMeasFromXY(0, 3));

        List<? extends Kaly2Feature> features = merge.getFeatures(measurements);

        assertEquals(features.get(0).getY(), 0, EPSILON);
        assertEquals(features.get(1).getY(), 3, EPSILON);
        assertTrue(features.size() == 2);
    }

    @Test
    public void testInnerCorner() {
        SplitAndMerge merge = new SplitAndMerge();

        ArrayList<Measurement> measurements = new ArrayList<>();
        for (float i = 0; i <= 1; i += 0.1) {
            measurements.add(makeMeasFromXY(1, i));
        }
        for (float i = 0; i <= 1; i += 0.1) {
            measurements.add(makeMeasFromXY(i, 1));
        }

        List<? extends Kaly2Feature> features = merge.getFeatures(measurements);

        assertTrue(features.size() == 3);
    }

    @Test
    public void testOuterCorner() {
        SplitAndMerge merge = new SplitAndMerge();

        ArrayList<Measurement> measurements = new ArrayList<>();
        for (float i = 1; i >= 0; i -= 0.1) {
            measurements.add(makeMeasFromXY(0, i));
        }
        for (float i = 1; i >= 0; i -= 0.1) {
            measurements.add(makeMeasFromXY(i, 0));
        }

        List<? extends Kaly2Feature> features = merge.getFeatures(measurements);

        assertTrue(features.size() == 3);
    }

  /*
  This test will work once the full split-and-merge is implemented
  @Test
  public void testDenseSquare() {
    SplitAndMerge merge = new SplitAndMerge();

    ArrayList<Measurement> measurements = new ArrayList<>();
    for(float i = -1; i<= 1; i+= 0.1) {
      measurements.add(makeMeasFromXY(-1,i));
    }
    for(float i = -1; i<= 1; i+= 0.1) {
      measurements.add(makeMeasFromXY(i,1));
    }
    for(float i = 1; i >= -1; i-= 0.1) {
      measurements.add(makeMeasFromXY(1,i));
    }
    for(float i = 1; i >= -1; i-= 0.1) {
      measurements.add(makeMeasFromXY(i,-1));
    }

    List<? extends Kaly2Feature> features = merge.getFeatures(measurements);

    assertTrue(features.size() == 4);
  }
  */
}

File: src/test/kotlin/ca/joelathiessen/kaly2/tests/pc/unit/featuredetector/SplitAndMergeTest.kt
package ca.joelathiessen.kaly2.tests.pc.unit.featuredetector

import ca.joelathiessen.kaly2.Measurement
import ca.joelathiessen.kaly2.featuredetector.SplitAndMerge
import ca.joelathiessen.kaly2.tests.pc.unit.util.*
import org.junit.Assert.assertEquals
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.runners.MockitoJUnitRunner
import java.util.*

@RunWith(MockitoJUnitRunner::class)
class SplitAndMergeTest {
    private val MAX_ERROR = 0.001f
    private val LINE_THRESHOLD = 0.1f
    private val CHECK_WITHIN_ANGLE = 0.3f
    private val MAX_RATIO = 1.0f
    private val ROBOT_POS = 0.5f

    @Test
    fun testFourPointSquare() {
        val merge = SplitAndMerge(LINE_THRESHOLD, CHECK_WITHIN_ANGLE, MAX_RATIO)

        val measurements = ArrayList<Measurement>()
        measurements.add(makeMeasFromXY(0.0f, 0.0f, ROBOT_POS, ROBOT_POS))
        measurements.add(makeMeasFromXY(0.0f, 1.0f, ROBOT_POS, ROBOT_POS))
        measurements.add(makeMeasFromXY(1.0f, 1.0f, ROBOT_POS, ROBOT_POS))
        measurements.add(makeMeasFromXY(1.0f, 0.0f, ROBOT_POS, ROBOT_POS))

        val features = merge.getFeatures(measurements)

        assertEquals(features.size, measurements.size)

        assertEquals(features[0].x, 0.0f, MAX_ERROR)
        assertEquals(features[0].y, 0.0f, MAX_ERROR)

        assertEquals(features[1].x, 1.0f, MAX_ERROR)
        assertEquals(features[1].y, 0.0f, MAX_ERROR)

        assertEquals(features[2].x, 1.0f, MAX_ERROR)
        assertEquals(features[2].y, 1.0f, MAX_ERROR)

        assertEquals(features[3].x, 0.0f, MAX_ERROR)
        assertEquals(features[3].y, 1.0f, MAX_ERROR)
    }

    @Test
    fun testFourPointLine() {
        val merge = SplitAndMerge(LINE_THRESHOLD, CHECK_WITHIN_ANGLE, MAX_RATIO)

        val measurements = ArrayList<Measurement>()
        measurements.add(makeMeasFromXY(0.0f, 0.0f, ROBOT_POS, ROBOT_POS))
        measurements.add(makeMeasFromXY(0.0f, 1.0f, ROBOT_POS, ROBOT_POS))
        measurements.add(makeMeasFromXY(0.0f, 2.0f, ROBOT_POS, ROBOT_POS))
        measurements.add(makeMeasFromXY(0.0f, 3.0f, ROBOT_POS, ROBOT_POS))

        val features = merge.getFeatures(measurements)

        assertEquals(features.size, 0) // all features within the threshold should are discarded
    }

    @Test
    fun testInnerCorner() {
        val merge = SplitAndMerge(LINE_THRESHOLD, CHECK_WITHIN_ANGLE, MAX_RATIO)

        val measurements = ArrayList<Measurement>()
        run {
            var i = 0.0f
            while (i <= 1f) {
                measurements.add(makeMeasFromXY(1.0f, i, ROBOT_POS, ROBOT_POS))
                i += 0.1f
            }
        }
        var i = 0.0f
        while (i <= 1f) {
            measurements.add(makeMeasFromXY(i, 1.0f, ROBOT_POS, ROBOT_POS))
            i += 0.1f
        }

        val features = merge.getFeatures(measurements)

        assertEquals(features.size, 3)
    }

    @Test
    fun testOuterCorner() {
        val merge = SplitAndMerge(LINE_THRESHOLD, CHECK_WITHIN_ANGLE, MAX_RATIO)

        val measurements = ArrayList<Measurement>()
        run {
            var i = 1.0f
            while (i >= 0f) {
                measurements.add(makeMeasFromXY(0.0f, i, ROBOT_POS, ROBOT_POS))
                i -= 0.1f
            }
        }
        var i = 1.0f
        while (i >= 0f) {
            measurements.add(makeMeasFromXY(i, 0.0f, ROBOT_POS, ROBOT_POS))
            i -= 0.1f
        }

        val features = merge.getFeatures(measurements)

        assertEquals(features.size, 3)
    }

    @Test
    fun testDenseSquare() {
        val merge = SplitAndMerge(LINE_THRESHOLD, CHECK_WITHIN_ANGLE, MAX_RATIO)
        val measurements = ArrayList<Measurement>()

        var left = -1.0f
        while (left < 1.0f) {
            measurements.add(makeMeasFromXY(-1.0f, left, ROBOT_POS, ROBOT_POS))
            left += 0.01f
        }

        var top = -1.0f
        while (top < 1.0f) {
            measurements.add(makeMeasFromXY(top, 1.0f, ROBOT_POS, ROBOT_POS))
            top += 0.01f
        }

        var right = 1.0f
        while (right > -1.0f) {
            measurements.add(makeMeasFromXY(1.0f, right, ROBOT_POS, ROBOT_POS))
            right -= 0.01f
        }

        var bottom = 1.0f
        while (bottom > -1.0f) {
            measurements.add(makeMeasFromXY(bottom, -1.0f, ROBOT_POS, ROBOT_POS))
            bottom -= 0.01f
        }

        val features = merge.getFeatures(measurements)

        assertEquals(features.size, 4)

        assertEquals(features[0].x, -1.0f, MAX_ERROR)
        assertEquals(features[0].y, -1.0f, MAX_ERROR)

        assertEquals(features[1].x, 1.0f, MAX_ERROR)
        assertEquals(features[1].y, -1.0f, MAX_ERROR)

        assertEquals(features[2].x, 1.0f, MAX_ERROR)
        assertEquals(features[2].y, 1.0f, MAX_ERROR)

        assertEquals(features[3].x, -1.0f, MAX_ERROR)
        assertEquals(features[3].y, 1.0f, MAX_ERROR)
    }
}
